### DISCUSSION TITLE
einops for C++ frontend

### DISCUSSION THREAD

**Discussion by pfeatherstone** [arogozhnikov/einops#226] [Ideas]
It would be really cool if there was a version of einops for the C++ frontend.
I imagine you could do cool things like parsing the expressions at compile time using something like [ctre](https://github.com/hanickadot/compile-time-regular-expressions). But that's an implementation detail.
I have a use-case for training in c++ and could really benefit with einops

**Comment by ghost** [comment]
I've do a basic and partial implementation for my personal usage, maybe that can help : https://github.com/dorpxam/einops-cpp

**Reply by arogozhnikov** [reply]
Nice project, I'll add this to the list of related projects

**Reply by ghost** [reply]
I am currently working on a more successful port (finalization of existing methods plus einsum support and more) with potentially the integration of tensorflow in the backends, and better compatibility in pure C++20. I will update the repo in few days.

**Comment by dorpxam** [comment]
Hi. Sorry for the 'ghost' but I just restart my old github (not really used) to a fresh one. More clean and dedicated to future (more professional) projects.

The URL is now : https://github.com/dorpxam/einops-cpp (not the old https://github.com/MaxCoo/einops-cpp)

I've rewrite the code to follow (when possible) the latest version v0.7, what's the new?

- The code is now C++17 because C++20 is too immature for a large compiler support.
- The code is now free of dependencies, I integrate now the python-like iterators (no more CPPIterTools)
- The code is always header-only, but I've add a Cmake project for building the tests.
- The code is now complete for reduce(), rearrange(), repeat(), einsum(), parse_shape().
- The code integrate now a LRU Cache (need to be benchmarked)

Unfortunatly, no more backend than the Libtorch one for this version. I explain why in the 'Future version and refactoring' section of the README.md.

I hope you'll appreciate this new version.

Cheers,
Max

**Reply by arogozhnikov** [reply]
Cool, for the future - you could just move repo between accounts. It would preserve git origins, history, likes, etc. Not a big deal, but convenient.

**Reply by dorpxam** [reply]
This is my first github project, and very newbie with git, cmake ... This repo will stay in place now. But thank's for the info !
